text/trace/0189-instrumentation-layers-suppression.md

```diff
@@ -0,0 +1,127 @@
+# Instrumentation Layers and Suppression
+
+This document describes approach for instrumentation layers, suppressing duplicate layers and unambiguously enriching spans.
+
+## Motivation
+
+- Provide clarity for instrumentation layers: e.g. DB calls on top of REST API
+- Give mechanism to suppress instrumentation layers for the same type: e.g. multiple instrumented HTTP clients using each other.
+- Give mechanism to enrich specific spans unambiguously: e.g. HTTP server span with routing information
+
+## Explanation
+
+### Spec changes proposal
+
+- Semantic Conventions: Each span MUST follow one convention, specific to the call it describes
+- Trace API: Add `SpanKey` API that gets span following specific convention from the context (e.g. `SpanKey.HTTP_CLIENT.fromContextOrNull(context)`).
+- Semantic Conventions: instrumentation MUST back off if span of same kind and following same contention is already in the context by using `ContextKey` API.

```

Comment:
I honestly don't know. We do have clients who want to have as much information as possible.

Do you think that having this "no-suppression" option will complicate the design considerably? If not, I think we should not take the option away from the clients. We cannot predict the future :)

If this is actually complicated, then maybe having lower non-http layer is actually enough...

Reply:
Yes, there's no additional complexity with adding no-suppression option. I think it can be useful for debugging instrumentation issues. I'll add it to the doc later this week and hope we'll get a chance to discuss it on one of the instrumentation sigs